Univeral not null calculation; Input format autodetection; Targets dataset

tskir opened the pull request:
* Universal not null calculation
* Rewrote the input format autodetection and the `--evidence` handling
* Amended the list of system level dependencies and added suggestions for running in Google Cloud.
* Added the targets dataset

As we discussed during the stand-up, there's no need in the distinct count calculation for anything other than the evidence datasets.

ireneisdoomed commented on metrics.py:
```diff
@@ -67,41 +67,40 @@ def document_count_by(
-                  value_vars=cols,
-                  value_name='count')
+    # Get the list of fields to count.
+    field_list = flat_df.schema
```
```suggestion
    field_list = flat_df.columns if not group_by_datasource else flat_df.drop('datasourceId').columns
```

This would remove lines 82-84. Also, do you use `.schema` because you need a list of the datatypes? If you simply need a list of the columns, .`columns` works just like in Pandas :)

tskir replied:
Oh, you're absolutely right! The use of `.schema` was there before, and I just refactored it to be the same without thinking. This is indeed much more elegant; addressed in https://github.com/opentargets/ot-release-metrics/pull/4/commits/cadd7a0ad8ce1bf4c090a78fc364e399de254bc7.

tskir replied:
Whoops, actually no, this breaks things. In calculating `exprs`, we check `if isinstance(field.dataType, t.ArrayType)`, so we can't have the field be just a string. Sorry, I also didn't notice this. Reverted in https://github.com/opentargets/ot-release-metrics/pull/4/commits/73a4bc8a583bcb0f591cf32f6fea631017161136

ireneisdoomed commented on metrics.py:
```diff
@@ -67,41 +67,40 @@ def document_count_by(
+    # Clean column names.
+    df_cleaned = df_aggregated.toDF(*(c.replace('.', '_') for c in df_aggregated.columns))
+    # Wide to long format.
+    cols = df_cleaned.drop('datasourceId').columns if group_by_datasource else df_cleaned.columns
```
This variable could be merged with `field_list`, right?

tskir replied:
Yep, you're right. Another case of refactoring without thinking too much :man_facepalming: Addressed in https://github.com/opentargets/ot-release-metrics/pull/4/commits/cadd7a0ad8ce1bf4c090a78fc364e399de254bc7.

tskir replied:
Same as for [this thread](https://github.com/opentargets/ot-release-metrics/pull/4#discussion_r620498335), this was actually required: we have renamed some columns in the dataframe already, and now we do need to re-calculate their names. Reverted in https://github.com/opentargets/ot-release-metrics/pull/4/commits/73a4bc8a583bcb0f591cf32f6fea631017161136

ireneisdoomed commented on metrics.py:
```diff
@@ -67,41 +67,40 @@ def document_count_by(
+    df_cleaned = df_aggregated.toDF(*(c.replace('.', '_') for c in df_aggregated.columns))
+    # Wide to long format.
+    cols = df_cleaned.drop('datasourceId').columns if group_by_datasource else df_cleaned.columns
+    melted = melt(df_cleaned, id_vars=['datasourceId'] if group_by_datasource else [], var_name='field',
```
```suggestion
    melted = (
        melt(df_cleaned,
            id_vars=['datasourceId'] if group_by_datasource else [],
            var_name='field',
            value_vars=cols,
            value_name='count')
        .withColumn('variable', f.lit(var_name)))
```

tskir replied:
Thank you for pointing this out. I've reformatted this slightly differently, but along the same lines: https://github.com/opentargets/ot-release-metrics/pull/4/commits/874e7a4b88b8f9d36c2a2772dfe1780c6a301a5d, do you like it?

ireneisdoomed commented on metrics.py:
```diff
@@ -67,41 +67,40 @@ def document_count_by(
-        var_name: str) -> DataFrame:
-    """Count number of evidences with not null values in variable."""
-    # flatten dataframe schema
+def not_null_fields_count(df: DataFrame, var_name: str, group_by_datasource: bool) -> DataFrame:
```
To follow the same style as the other functions.

```suggestion
def not_null_fields_count(
        df: DataFrame,
        var_name: str,
        group_by_datasource: bool) -> DataFrame:
```

tskir replied:
Good point. I've looked online a bit, and I found that this style is often used for signatures with type hints; almost identical, but the return value is also on a newline:

```python
def not_null_fields_count(
        df: DataFrame,
        var_name: str,
        group_by_datasource: bool
) -> DataFrame:
```

I've reformatted all functions with type hints in https://github.com/opentargets/ot-release-metrics/pull/4/commits/9c2d36e068315212078aef139e19faf4e4d2fffc; do you like it?

ireneisdoomed commented on metrics.py:
```diff
@@ -67,41 +67,40 @@ def document_count_by(
-    """Count number of evidences with not null values in variable."""
-    # flatten dataframe schema
+def not_null_fields_count(df: DataFrame, var_name: str, group_by_datasource: bool) -> DataFrame:
+    """Count number of not null values for each field in the dataframe. If `group_by_column` is provided, the
```
Change `group_by_column` to `group_by_datasource`

tskir replied:
Good catch, fixed in https://github.com/opentargets/ot-release-metrics/pull/4/commits/cadd7a0ad8ce1bf4c090a78fc364e399de254bc7.

ireneisdoomed commented on metrics.py:
```diff
@@ -216,33 +226,27 @@ def main(args):
 
     if evidence:
         logging.info(f'Running metrics from {evidence_filename}.')
+        columns_to_report = get_columns_to_report(evidence.columns)
```
Nice function! We can expand it to report columns present in different datasets.

tskir replied:
Thank you ^__^

> We can expand it to report columns present in different datasets.

Do you mean right away, or for subsequent tickets/PRs?

ireneisdoomed replied:
Subsequent, do not worry :)

ireneisdoomed reviewed (changes requested):
Great work detangling the not null function :) Let me know how you feel about the comments.

tskir commented:
@ireneisdoomed Thank you for the review! Addressed your comments and implemented the single `--evidence` parameter with improved input format autodetection (https://github.com/opentargets/ot-release-metrics/pull/4/commits/3cadfb90e479a93c91b854dd5d050a5618e2d3e1)

tskir commented:
@ireneisdoomed Also added the targets dataset as discussed: https://github.com/opentargets/ot-release-metrics/pull/4/commits/c1119288d88b2402ccb5ef75b81043c370c6e245.

I've also removed the TODOs from diseases and drugs, as those are now implemented.

ireneisdoomed reviewed (approved):
Everything looks great, thanks for addressing all the comments!!

DSuveges reviewed (approved):
All looks good.
